File: quotio-cli/src/services/tunnel/cloudflared-service.ts
import { type ChildProcess, spawn } from "node:child_process";
import { execSync } from "node:child_process";
import { existsSync } from "node:fs";

export type TunnelStatus =
	| "idle"
	| "starting"
	| "active"
	| "stopping"
	| "error";

export interface TunnelState {
	status: TunnelStatus;
	publicURL: string | null;
	errorMessage: string | null;
	startTime: string | null;
}

export interface CloudflaredInstallation {
	isInstalled: boolean;
	path: string | null;
	version: string | null;
}

const TUNNEL_URL_PATTERN = /https:\/\/[a-z0-9-]+\.trycloudflare\.com/;

function getBinaryPaths(): string[] {
	if (process.platform === "win32") {
		return [
			"C:\\Program Files\\cloudflared\\cloudflared.exe",
			"C:\\Program Files (x86)\\cloudflared\\cloudflared.exe",
			`${process.env.LOCALAPPDATA}\\cloudflared\\cloudflared.exe`,
		];
	}
	if (process.platform === "darwin") {
		return [
			"/opt/homebrew/bin/cloudflared",
			"/usr/local/bin/cloudflared",
			"/usr/bin/cloudflared",
		];
	}
	return [
		"/usr/local/bin/cloudflared",
		"/usr/bin/cloudflared",
		"/snap/bin/cloudflared",
	];
}

function detectInstallation(): CloudflaredInstallation {
	for (const path of getBinaryPaths()) {
		if (existsSync(path)) {
			const version = getVersion(path);
			return { isInstalled: true, path, version };
		}
	}
	return { isInstalled: false, path: null, version: null };
}

function getVersion(binaryPath: string): string | null {
	try {
		const output = execSync(`"${binaryPath}" --version`, {
			encoding: "utf-8",
			timeout: 5000,
		});
		const match = output.match(/\d+\.\d+\.\d+/);
		return match ? match[0] : null;
	} catch {
		return null;
	}
}

class CloudflaredService {
	private process: ChildProcess | null = null;
	private state: TunnelState = {
		status: "idle",
		publicURL: null,
		errorMessage: null,
		startTime: null,
	};
	private outputBuffer = "";
	private urlDetected = false;
	private startTimeoutHandle: ReturnType<typeof setTimeout> | null = null;

	getState(): TunnelState {
		return { ...this.state };
	}

	getInstallation(): CloudflaredInstallation {
		return detectInstallation();
	}

	async start(port: number): Promise<{ success: boolean; error?: string }> {
		if (this.process) {
			return { success: false, error: "Tunnel is already running" };
		}

		const installation = detectInstallation();
		if (!installation.isInstalled || !installation.path) {
			this.state = {
				status: "error",
				publicURL: null,
				errorMessage: "Cloudflared is not installed",
				startTime: null,
			};
			return { success: false, error: "Cloudflared is not installed" };
		}

		const binaryPath = installation.path;
		this.state = {
			status: "starting",
			publicURL: null,
			errorMessage: null,
			startTime: null,
		};
		this.outputBuffer = "";
		this.urlDetected = false;

		return new Promise((resolve) => {
			const args = [
				"tunnel",
				"--config",
				process.platform === "win32" ? "NUL" : "/dev/null",
				"--url",
				`http://localhost:${port}`,
			];

			try {
				this.process = spawn(binaryPath, args, {
					stdio: ["ignore", "pipe", "pipe"],
				});
			} catch (err) {
				const error = err instanceof Error ? err.message : String(err);
				this.state = {
					status: "error",
					publicURL: null,
					errorMessage: error,
					startTime: null,
				};
				this.cleanup();
				resolve({ success: false, error });
				return;
			}

			this.startTimeoutHandle = setTimeout(() => {
				if (this.state.status === "starting") {
					this.state = {
						status: "error",
						publicURL: null,
						errorMessage: "Tunnel start timed out",
						startTime: null,
					};
					this.stop();
				}
			}, 30000);

			const handleOutput = (data: Buffer) => {
				if (this.urlDetected) return;
				this.outputBuffer += data.toString("utf-8");
				if (this.outputBuffer.length > 65536) {
					this.outputBuffer = this.outputBuffer.slice(-65536);
				}
				const match = this.outputBuffer.match(TUNNEL_URL_PATTERN);
				if (match) {
					this.urlDetected = true;
					this.state = {
						status: "active",
						publicURL: match[0],
						errorMessage: null,
						startTime: new Date().toISOString(),
					};
					this.clearStartTimeout();
				}
			};

			this.process.stdout?.on("data", handleOutput);
			this.process.stderr?.on("data", handleOutput);

			this.process.on("error", (err) => {
				this.state = {
					status: "error",
					publicURL: null,
					errorMessage: err.message,
					startTime: null,
				};
				this.cleanup();
			});

			this.process.on("exit", (code) => {
				if (
					this.state.status === "active" ||
					this.state.status === "starting"
				) {
					this.state = {
						status: "error",
						publicURL: null,
						errorMessage: `Tunnel exited unexpectedly (code: ${code})`,
						startTime: null,
					};
				}
				this.cleanup();
			});

			setTimeout(() => {
				if (this.process && !this.process.killed) {
					resolve({ success: true });
				}
			}, 500);
		});
	}

	async stop(): Promise<void> {
		this.clearStartTimeout();
		if (!this.process) {
			this.state = {
				status: "idle",
				publicURL: null,
				errorMessage: null,
				startTime: null,
			};
			return;
		}

		this.state = { ...this.state, status: "stopping" };

		const proc = this.process;
		proc.kill("SIGTERM");

		await new Promise<void>((resolve) => {
			const timeout = setTimeout(() => {
				if (!proc.killed) {
					proc.kill("SIGKILL");
				}
				resolve();
			}, 500);

			proc.once("exit", () => {
				clearTimeout(timeout);
				resolve();
			});
		});

		this.cleanup();
		this.state = {
			status: "idle",
			publicURL: null,
			errorMessage: null,
			startTime: null,
		};
	}

	isRunning(): boolean {
		return this.process !== null && !this.process.killed;
	}

	private cleanup(): void {
		this.clearStartTimeout();
		if (this.process) {
			this.process.stdout?.removeAllListeners();
			this.process.stderr?.removeAllListeners();
			this.process.removeAllListeners();
			this.process = null;
		}
		this.outputBuffer = "";
	}

	private clearStartTimeout(): void {
		if (this.startTimeoutHandle) {
			clearTimeout(this.startTimeoutHandle);
			this.startTimeoutHandle = null;
		}
	}
}

let serviceInstance: CloudflaredService | null = null;

export function getCloudflaredService(): CloudflaredService {
	if (!serviceInstance) {
		serviceInstance = new CloudflaredService();
	}
	return serviceInstance;
}

File: Quotio/Services/Tunnel/CloudflaredService.swift
//
//  CloudflaredService.swift
//  Quotio - Cloudflared subprocess management
//

import Foundation

actor CloudflaredService {
    
    private var process: Process?
    private var outputPipe: Pipe?
    private var errorPipe: Pipe?
    
    private static let binaryPaths = [
        "/opt/homebrew/bin/cloudflared",
        "/usr/local/bin/cloudflared",
        "/usr/bin/cloudflared"
    ]
    
    private static let httpsURLPattern = #"https://[A-Za-z0-9._~:/?#\[\]@!\$&'()*+,;=%-]+"#
    
    nonisolated func detectInstallation() -> CloudflaredInstallation {
        for path in Self.binaryPaths {
            if FileManager.default.isExecutableFile(atPath: path) {
                let version = getVersion(at: path)
                return CloudflaredInstallation(isInstalled: true, path: path, version: version)
            }
        }
        return .notInstalled
    }
    
    private nonisolated func getVersion(at path: String) -> String? {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: path)
        process.arguments = ["--version"]
        
        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = FileHandle.nullDevice
        
        do {
            try process.run()
            process.waitUntilExit()
            
            let data = pipe.fileHandleForReading.readDataToEndOfFile()
            guard let output = String(data: data, encoding: .utf8) else { return nil }
            
            if let match = output.range(of: #"\d+\.\d+\.\d+"#, options: .regularExpression) {
                return String(output[match])
            }
            return nil
        } catch {
            return nil
        }
    }

    nonisolated private static func extractPublicURL(from text: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: httpsURLPattern) else { return nil }
        let nsText = text as NSString
        let matches = regex.matches(in: text, range: NSRange(location: 0, length: nsText.length))

        for match in matches {
            guard match.range.location != NSNotFound else { continue }
            let candidate = nsText.substring(with: match.range)
            guard let url = URL(string: candidate), let host = url.host?.lowercased() else { continue }

            // Keep the quick tunnel URL and allow custom domains.
            if host.hasSuffix("trycloudflare.com") {
                return candidate
            }

            // Skip Cloudflare control-plane URLs; keep user-owned hostnames.
            if host.hasSuffix("cloudflare.com") || host.hasSuffix("cloudflareclient.com") {
                continue
            }

            return candidate
        }

        return nil
    }
    
    func start(
        port: UInt16,
        onURLDetected: @escaping @Sendable (String) -> Void,
        onLogLine: (@Sendable (String, Bool) -> Void)? = nil
    ) async throws {
        guard process == nil else {
            throw TunnelError.alreadyRunning
        }
        
        let installation = detectInstallation()
        guard installation.isInstalled, let binaryPath = installation.path else {
            throw TunnelError.notInstalled
        }
        
        let newProcess = Process()
        newProcess.executableURL = URL(fileURLWithPath: binaryPath)

        let tunnelToken = await MainActor.run {
            KeychainHelper.getCloudflareTunnelToken()?.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if let token = tunnelToken, !token.isEmpty {
            // Token mode uses named tunnels configured in Cloudflare dashboard.
            // Include --url so origin always matches the current local proxy port.
            newProcess.arguments = ["tunnel", "run", "--token", token, "--url", "http://localhost:" + String(port)]
        } else {
            // Use --config /dev/null to ignore user's existing config file.
            // This ensures Quick Tunnel works without interference from named tunnels.
            newProcess.arguments = ["tunnel", "--config", "/dev/null", "--protocol", "http2", "--url", "http://localhost:" + String(port)]
        }
        
        let outputPipe = Pipe()
        let errorPipe = Pipe()
        newProcess.standardOutput = outputPipe
        newProcess.standardError = errorPipe
        
        self.outputPipe = outputPipe
        self.errorPipe = errorPipe
        
        final class OutputBuffer: @unchecked Sendable {
            private let lock = NSLock()
            private var buffer = ""
            private var urlFound = false
            private let maxBufferSize = 65536 // 64KB max buffer
            
            func append(_ text: String) -> String? {
                lock.lock()
                defer { lock.unlock() }
                
                guard !urlFound else { return nil }
                buffer += text
                
                // Trim buffer to keep only trailing maxBufferSize characters
                if buffer.count > maxBufferSize {
                    let dropCount = buffer.count - maxBufferSize
                    buffer = String(buffer.dropFirst(dropCount))
                }
                
                if let url = CloudflaredService.extractPublicURL(from: buffer) {
                    urlFound = true
                    return url
                }
                return nil
            }
            
            func checkRemaining() -> String? {
                lock.lock()
                defer { lock.unlock() }
                
                guard !urlFound else { return nil }
                
                if let url = CloudflaredService.extractPublicURL(from: buffer) {
                    urlFound = true
                    return url
                }
                return nil
            }
        }

        final class LineBuffer: @unchecked Sendable {
            private let lock = NSLock()
            private var pending = ""

            func append(_ text: String) -> [String] {
                lock.lock()
                defer { lock.unlock() }

                pending += text.replacingOccurrences(of: "\r\n", with: "\n").replacingOccurrences(of: "\r", with: "\n")
                let parts = pending.components(separatedBy: "\n")
                pending = parts.last ?? ""
                guard parts.count > 1 else { return [] }
                return parts[0..<(parts.count - 1)].filter { !$0.isEmpty }
            }

            func flush() -> [String] {
                lock.lock()
                defer { lock.unlock() }

                let last = pending.trimmingCharacters(in: .whitespacesAndNewlines)
                pending = ""
                return last.isEmpty ? [] : [last]
            }
        }
        
        let buffer = OutputBuffer()
        let stdoutLines = LineBuffer()
        let stderrLines = LineBuffer()
        
        let emitLogLines: @Sendable (String, Bool) -> Void = { text, isErrorStream in
            guard let onLogLine else { return }
            let lineBuffer = isErrorStream ? stderrLines : stdoutLines
            let lines = lineBuffer.append(text)
            for line in lines {
                onLogLine(line, isErrorStream)
            }
        }

        let emitRemainingLines: @Sendable (Bool) -> Void = { isErrorStream in
            guard let onLogLine else { return }
            let lineBuffer = isErrorStream ? stderrLines : stdoutLines
            let lines = lineBuffer.flush()
            for line in lines {
                onLogLine(line, isErrorStream)
            }
        }

        errorPipe.fileHandleForReading.readabilityHandler = { handle in
            let data = handle.availableData
            
            // EOF detected - empty data means stream closed
            if data.isEmpty {
                handle.readabilityHandler = nil
                emitRemainingLines(true)
                // Check remaining buffer for URL on EOF
                if let url = buffer.checkRemaining() {
                    onURLDetected(url)
                }
                return
            }
            
            guard let text = String(data: data, encoding: .utf8) else { return }
            emitLogLines(text, true)
            
            if let url = buffer.append(text) {
                onURLDetected(url)
            }
        }
        
        outputPipe.fileHandleForReading.readabilityHandler = { handle in
            let data = handle.availableData
            
            // EOF detected - empty data means stream closed
            if data.isEmpty {
                handle.readabilityHandler = nil
                emitRemainingLines(false)
                // Check remaining buffer for URL on EOF
                if let url = buffer.checkRemaining() {
                    onURLDetected(url)
                }
                return
            }
            
            guard let text = String(data: data, encoding: .utf8) else { return }
            emitLogLines(text, false)
            
            if let url = buffer.append(text) {
                onURLDetected(url)
            }
        }
        
        do {
            try newProcess.run()
            self.process = newProcess
            NSLog("[CloudflaredService] Started tunnel on port %d, PID: %d", port, newProcess.processIdentifier)
        } catch {
            cleanup()
            throw TunnelError.startFailed(error.localizedDescription)
        }
    }
    
    func stop() async {
        guard let process = process, process.isRunning else {
            cleanup()
            return
        }
        
        let pid = process.processIdentifier
        NSLog("[CloudflaredService] Stopping tunnel, PID: %d", pid)
        
        process.terminate()
        
        let deadline = Date().addingTimeInterval(0.5)
        while process.isRunning && Date() < deadline {
            try? await Task.sleep(nanoseconds: 50_000_000)
        }
        
        if process.isRunning {
            NSLog("[CloudflaredService] Force killing tunnel, PID: %d", pid)
            kill(pid, SIGKILL)
        }
        
        cleanup()
    }
    
    var isRunning: Bool {
        process?.isRunning ?? false
    }
    
    private func cleanup() {
        outputPipe?.fileHandleForReading.readabilityHandler = nil
        errorPipe?.fileHandleForReading.readabilityHandler = nil
        
        try? outputPipe?.fileHandleForReading.close()
        try? errorPipe?.fileHandleForReading.close()
        
        outputPipe = nil
        errorPipe = nil
        process = nil
    }
    
    nonisolated static func killOrphanProcesses() {
        // Match both quick-tunnel and token-mode processes spawned by Quotio.
        let patterns = [
            "cloudflared.*tunnel.*--config.*/dev/null.*--url",
            "cloudflared.*tunnel.*run.*--token"
        ]

        for pattern in patterns {
            killProcesses(matching: pattern)
        }
    }

    nonisolated private static func killProcesses(matching pattern: String) {
        let termProcess = Process()
        termProcess.executableURL = URL(fileURLWithPath: "/usr/bin/pkill")
        termProcess.arguments = ["-TERM", "-f", pattern]
        termProcess.standardOutput = FileHandle.nullDevice
        termProcess.standardError = FileHandle.nullDevice

        do {
            try termProcess.run()
            termProcess.waitUntilExit()

            // Wait briefly for graceful shutdown.
            Thread.sleep(forTimeInterval: 0.3)

            let killProcess = Process()
            killProcess.executableURL = URL(fileURLWithPath: "/usr/bin/pkill")
            killProcess.arguments = ["-9", "-f", pattern]
            killProcess.standardOutput = FileHandle.nullDevice
            killProcess.standardError = FileHandle.nullDevice

            try killProcess.run()
            killProcess.waitUntilExit()
            NSLog("[CloudflaredService] Cleaned up orphan cloudflared processes for pattern: %@", pattern)
        } catch {
            // Silent failure - no orphans to kill is fine.
        }
    }
}

enum TunnelError: Error, Sendable {
    case notInstalled
    case alreadyRunning
    case startFailed(String)
    case unexpectedExit
    
    var localizedMessage: String {
        switch self {
        case .notInstalled:
            return "Cloudflared is not installed"
        case .alreadyRunning:
            return "Tunnel is already running"
        case .startFailed(let reason):
            return "Failed to start tunnel: \(reason)"
        case .unexpectedExit:
            return "Tunnel exited unexpectedly"
        }
    }
}
